tests/unit/s2n_self_talk_inet_socket_test.c

```diff
@@ -0,0 +1,163 @@
+            s2n_connection_ptr_free);
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key * chain_and_key, s2n_cert_chain_and_key_ptr_free);
+    RESULT_GUARD_POSIX(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Setup config */
+    RESULT_GUARD_POSIX(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+    RESULT_GUARD_POSIX(s2n_connection_set_fd(client_conn, fd));
+    RESULT_GUARD_POSIX(s2n_config_set_cipher_preferences(config, "default"));
+    RESULT_GUARD_POSIX(s2n_config_set_unsafe_for_testing(config));
+    RESULT_GUARD_POSIX(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+    RESULT_GUARD_POSIX(s2n_connection_set_config(client_conn, config));
+
+    /* Do handshake */
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    RESULT_GUARD_POSIX(s2n_negotiate(client_conn, &blocked));
+    RESULT_ENSURE_EQ(client_conn->actual_protocol_version, S2N_TLS12);
+
+    char sync = 0;
+    char recv_buffer[1] = { 0 };
+
+    {
+        RESULT_GUARD_POSIX(read(read_pipe, &sync, 1));
+        RESULT_GUARD_POSIX(s2n_recv(client_conn, recv_buffer, 1, &blocked));
+        RESULT_ENSURE_EQ(memcmp(&CHAR_A, &recv_buffer[0], 1), 0);
+
+        RESULT_GUARD_POSIX(read(read_pipe, &sync, 1));
+        RESULT_GUARD_POSIX(s2n_recv(client_conn, recv_buffer, 1, &blocked));
+        RESULT_ENSURE_EQ(memcmp(&CHAR_B, &recv_buffer[0], 1), 0);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT start_server(int fd, int write_pipe)
+{
+    /* Setup connections */
+    DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+            s2n_connection_ptr_free);
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key * chain_and_key, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Setup config */
+    EXPECT_SUCCESS(s2n_connection_set_blinding(server_conn, S2N_SELF_SERVICE_BLINDING));
+    EXPECT_EQUAL(s2n_connection_get_delay(server_conn), 0);
+    EXPECT_SUCCESS(s2n_connection_set_fd(server_conn, fd));
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+    EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+    /* Do handshake */
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    EXPECT_SUCCESS(s2n_negotiate(server_conn, &blocked));
+    EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+
+    char sync = 0;
+    char send_buffer[1] = { 0 };
```

Comment:
```suggestion
    char send_buffer[] = { 0 };
```
Nit: In C you don't need to specify how large the buffer is if you're initializing it with elements.